src/plan/plan_constraints.rs

```diff
@@ -13,6 +14,7 @@ pub struct PlanConstraints {
     pub needs_write_barrier: bool,
     pub needs_log_bit_in_header: bool,
     pub needs_log_bit_in_header_num: usize,
+    pub barrier: BarrierSelector,
```

Comment:
Does this `barrier` refer to write barrier only? Should we merge `needs_write_barrier` with `barrier`? For example, remove `needs_write_barrier` - basically `BarrierSelector::NoBarrier` means `false` for `needs_write_barrier`.

Reply:
Should the `barrier` selector be consistent with the `barrier` field in `Mutator`? I would suggest adding a function `fn new_barrier(selector: BarrierSelector) -> Box<dyn Barrier>` to `barriers.rs` so that `Mutator.barrier` is created from the selector, and making constructors of the barrier implementations as private so that people can only create barrier from a selector. This could ensure the selector is always consistent with the actual barrier used.

Reply:
I've removed `needs_write_barrier`.

For the second suggestion, I think it is hard to use a factory method for all Barrier struct creation. Different barrier type requires different constructor arguments. Some barriers require a specific type of nursery space while some barriers do not require any arguments (e.g. NoBarrier).